File: app/src/main/java/ru/geekbrains/notes/repository/FirestoreNotesRepository.java
package ru.geekbrains.notes.repository;

import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.Date;
import java.util.HashMap;
import java.util.List;

import ru.geekbrains.notes.Note;

public class FirestoreNotesRepository {

    public FirestoreNotesRepository(String email) {
        NOTES_COLLECTION = email;
    }

    private final FirebaseFirestore FIRE_STORE = FirebaseFirestore.getInstance();

    private static String NOTES_COLLECTION = "notes";
    private final static String TITLE = "title";
    private final static String TEXT = "text";
    private final static String CREATED_AT = "createdAt";
    private final static String COLOR = "color";
    private final static String ID = "id";


    @RequiresApi(api = Build.VERSION_CODES.N)
    public void getNoteList(Callback<List<Note>> callback) {
        FIRE_STORE.collection(NOTES_COLLECTION)
                .get()
                .addOnCompleteListener(task -> {

                    if (task.isSuccessful()) {
                        ArrayList<Note> tmp = new ArrayList<>();

                        List<DocumentSnapshot> docs = task.getResult().getDocuments();
                        for (DocumentSnapshot doc : docs) {
                            String id = doc.getId();
                            String title = doc.getString(TITLE);
                            String text = doc.getString(TEXT);
                            Date date = doc.getDate(CREATED_AT);
                            int color = doc.getLong(COLOR).intValue();

                            tmp.add(new Note(id, title, text, color, date));
                        }

                        tmp.sort(Comparator.comparing(Note::getDate));
                        callback.onSuccess(tmp);


                    } else {
                        callback.onError(task.getException());
                    }
                });
    }

    public void addNote(Note note, Callback<Note> callback) {

        HashMap<String, Object> data = new HashMap<>();

        data.put(TITLE, note.getTitle());
        data.put(TEXT, note.getText());
        data.put(CREATED_AT, note.getDate());
        data.put(COLOR, note.getColor());
//        data.put(ID, note.getId());

        FIRE_STORE.collection(NOTES_COLLECTION)
                .add(data)
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        note.setId(task.getResult().getId());
                        callback.onSuccess(note);

                    } else {
                        callback.onError(task.getException());
                    }
                });
    }

    public void editNote(Note note, Callback<Note> callback) {

        HashMap<String, Object> data = new HashMap<>();

        data.put(TITLE, note.getTitle());
        data.put(TEXT, note.getText());
        data.put(CREATED_AT, note.getDate());
        data.put(COLOR, note.getColor());
        data.put(ID, note.getId());

        FIRE_STORE.collection(NOTES_COLLECTION)
                .document(note.getId())
                .set(data)
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        callback.onSuccess(note);

                    } else {
                        callback.onError(task.getException());
                    }
                });
    }

    public void removeNote(Note note, Callback<Note> callback) {

        FIRE_STORE.collection(NOTES_COLLECTION)
                .document(note.getId())
                .delete()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        callback.onSuccess(note);
                    } else {
                        callback.onError(task.getException());
                    }
                });
    }

    public void clearRepository(List<Note> noteList, Callback<Note> callback) {

        for (Note note : noteList) {
            FIRE_STORE.collection(NOTES_COLLECTION)
                    .document(note.getId())
                    .delete()
                    .addOnCompleteListener(task -> {
                        if (task.isSuccessful()) {
                            callback.onSuccess(note);
                        } else {
                            callback.onError(task.getException());
                        }
                    });
        }


    }




}

File: app/src/main/java/ru/geekbrains/notes/repository/FirestoreNotesRepository.kt
package ru.geekbrains.notes.repository

import com.google.firebase.firestore.FirebaseFirestore
import com.google.android.gms.tasks.Task
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.QuerySnapshot
import ru.geekbrains.notes.Note
import java.util.ArrayList
import java.util.Comparator
import java.util.HashMap

class FirestoreNotesRepository(email: String) {
    private val FIRE_STORE = FirebaseFirestore.getInstance()
    fun getNoteList(callback: Callback<MutableList<Note>>) {
        FIRE_STORE.collection(NOTES_COLLECTION)
            .get()
            .addOnCompleteListener { task: Task<QuerySnapshot> ->
                if (task.isSuccessful) {
                    val tmp = ArrayList<Note>()
                    val docs = task.result!!.documents
                    for (doc in docs) {
                        val id = doc.id
                        val title = doc.getString(TITLE)
                        val text = doc.getString(TEXT)
                        val date = doc.getDate(CREATED_AT)
                        val color = doc.getLong(COLOR)!!.toInt()
                        tmp.add(Note(id, title, text, color, date))
                    }
                    tmp.sortWith(Comparator.comparing { obj: Note -> obj.date })
                    callback.onSuccess(tmp)
                } else {
                    callback.onError(task.exception)
                }
            }
    }

    fun addNote(note: Note, callback: Callback<Note>) {
        val data = HashMap<String, Any>()
        data[TITLE] = note.title
        data[TEXT] = note.text
        data[CREATED_AT] = note.date
        data[COLOR] = note.color
        //        data.put(ID, note.getId());
        FIRE_STORE.collection(NOTES_COLLECTION)
            .add(data)
            .addOnCompleteListener { task: Task<DocumentReference> ->
                if (task.isSuccessful) {
                    note.id = task.result!!.id
                    callback.onSuccess(note)
                } else {
                    callback.onError(task.exception)
                }
            }
    }

    fun editNote(note: Note, callback: Callback<Note>) {
        val data = HashMap<String, Any>()
        data[TITLE] = note.title
        data[TEXT] = note.text
        data[CREATED_AT] = note.date
        data[COLOR] = note.color
        data[ID] = note.id
        FIRE_STORE.collection(NOTES_COLLECTION)
            .document(note.id)
            .set(data)
            .addOnCompleteListener { task: Task<Void?> ->
                if (task.isSuccessful) {
                    callback.onSuccess(note)
                } else {
                    callback.onError(task.exception)
                }
            }
    }

    fun removeNote(note: Note, callback: Callback<Note>) {
        FIRE_STORE.collection(NOTES_COLLECTION)
            .document(note.id)
            .delete()
            .addOnCompleteListener { task: Task<Void?> ->
                if (task.isSuccessful) {
                    callback.onSuccess(note)
                } else {
                    callback.onError(task.exception)
                }
            }
    }

    fun clearRepository(noteList: List<Note>, callback: Callback<Note>) {
        for (note in noteList) {
            FIRE_STORE.collection(NOTES_COLLECTION)
                .document(note.id)
                .delete()
                .addOnCompleteListener { task: Task<Void?> ->
                    if (task.isSuccessful) {
                        callback.onSuccess(note)
                    } else {
                        callback.onError(task.exception)
                    }
                }
        }
    }

    companion object {
        private var NOTES_COLLECTION = "notes"
        private const val TITLE = "title"
        private const val TEXT = "text"
        private const val CREATED_AT = "createdAt"
        private const val COLOR = "color"
        private const val ID = "id"
    }

    init {
        NOTES_COLLECTION = email
    }
}